cpp/src/arrow/type.h

```diff
@@ -719,7 +719,7 @@ class ARROW_EXPORT MapType : public ListType {
 };
 
 /// \brief Concrete type class for fixed size list data
-class ARROW_EXPORT FixedSizeListType : public NestedType {
+class ARROW_EXPORT FixedSizeListType : public BaseListType {
```

Comment:
I would check everywhere `enable_if_base_list`, `enable_if_fixed_size_list`, and `enable_if_list_like` are used. This might also create confusion between type traits name with `enable_if_base_binary` which `FixedSizedBinaryType` does not take part and why you have to use `enable_if_binary_like` (and `enable_if_list_like`).

So either removes this change, or remove `enable_if_list_like` and replace with usage of `enable_if_base_list`.

Reply:
Would introducing a listlike base class make sense?

Reply:
@fsaintjacques thinking about this some more I think the problem is that base_list BaseListArray is ambiguous.  I created a new type_tate is_var_length_list... to replace the template to be clearer. If this change sounds reasonable, I'll open up a JIRA and make a similar change for binary types.

The other options, per my previous comment, is to make a ListLikeBase and make BaseListType and  FixedSizeListType  inherit from it 
 without changing type traits.  I like this change less, since it introduces another layer in the class hierarchy which at the moment doesn't seem to serve a purpose (if we need functionality specified to variable width list types we should can introduce a new class BaseVariableWidthListType later for the same purpose).

A third option which I like even less is special casing logic for FixedListType in the parquet code because the classes don't share d a common ancestor (this just makes less conceptual sense to me).

Thoughts?

Reply:
@fsaintjacques ping?

Reply:
Sorry for the late answer. I agree with your general sentiment. I like your solution of `is_var_length_list` more than the other options. Those traits are used at callsites where we need to differentiate when the value size (length) can be found on a per-value basis or in the type instance.

The existing hierarchy and naming of type_traits is due to accumulate baggage, e.g. some code was using pattern matching with is_base_of and some newer code use the type_traits.